### PR TITLE
Add GeoPolygon and tests

### PR DESCRIPTION
**Pull request by zzzz-vincent**
This is the first commit for the new feature `Geo Filter by Polygon`(#795) that introduces the GeoPolygon struct, representing a polygon defined by a list of coordinates. It includes the check_point method to determine if a given point is inside the polygon. Test cases validate the functionality for different scenarios.

Summary:

- Added GeoPolygon struct with coordinate vertices.
- Implemented check_point method for point-in-polygon check.
- Added corresponding test cases for coverage.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo fmt` command prior to submission?
3. [x] Have you checked your code using `cargo clippy` command?

### CODE REVIEW

File: lib/segment/src/types.rs [hunk 1]
```diff
@@ -1196,6 +1196,32 @@ impl GeoRadius {
     }
 }
 
+/// Geo filter request
+///
+/// Matches coordinates inside the polygon, defined by the given coordinates in order.
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, PartialEq)]
```
Comment (agourlay) [inline lib/segment/src/types.rs:1202]: I propose to remove the `JsonSchema` derivation and introduce it with the API to remind us to update the schema.

File: lib/segment/src/types.rs [hunk 2]
```diff
@@ -1647,6 +1673,101 @@ mod tests {
...
+
+        let inside_result = radius.check_point(0.5, 0.5);
+        assert!(inside_result);
+
+        let outside_result = radius.check_point(1.5, 1.5);
+        assert!(!outside_result);
+    }
+
+    #[test]
+    fn test_geo_boundingbox_check_point() {
```
Comment (agourlay) [inline lib/segment/src/types.rs:1691]: Maybe we could give it a more precise name as it assess the edge case calculation over the prime meridian.

Comment (agourlay) [reply lib/segment/src/types.rs:1691]: discarding comment

File: lib/segment/src/types.rs [hunk 3]
```diff
@@ -1196,6 +1196,32 @@ impl GeoRadius {
...
+
+impl GeoPolygon {
+    pub fn check_point(&self, lon: f64, lat: f64) -> bool {
+        let polygon_points: Vec<Coord<f64>> = self
+            .points
+            .iter()
+            .map(|p| Coord { x: p.lon, y: p.lat })
+            .collect();
+
+        let polygon = Polygon::new(polygon_points.into(), vec![]);
```
Comment (agourlay) [inline lib/segment/src/types.rs:1217]: for other reviewers, the second argument here is a list of point to form an inner polygon (exterior + interior = donut like shape).

**Comment by agourlay** [comment]
@zzzz-vincent LGTM, nice self contained PR :+1: 
Just one tiny comment to address before shipping.

File: lib/segment/src/types.rs [hunk 4]
```diff
@@ -1196,6 +1196,32 @@ impl GeoRadius {
...
 }
 
+/// Geo filter request
+///
+/// Matches coordinates inside the polygon, defined by the given coordinates in order.
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, PartialEq)]
+#[serde(rename_all = "snake_case")]
+pub struct GeoPolygon {
+    /// Ordered list of coordinates representing the vertices of a polygon.
+    pub points: Vec<GeoPoint>,
```
Comment (agourlay) [inline lib/segment/src/types.rs:1206]: Something to keep in mind, why I do not want to expose right away this type on the API.

The GeoJSON specs says

```
   To specify a constraint specific to Polygons, it is useful to
   introduce the concept of a linear ring:

   o  A linear ring is a closed LineString with four or more positions.

   o  The first and last positions are equivalent, and they MUST contain
      identical values; their representation SHOULD also be identical.

   o  A linear ring is the boundary of a surface or the boundary of a
      hole in a surface.

   o  A linear ring MUST follow the right-hand rule with respect to the
      area it bounds, i.e., exterior rings are counterclockwise, and
      holes are clockwise.
```

This means we will have a validation layer to ensure those constraints are valid.

It seems the underlying geo crates handle both open & closed polygon transparently by closing the open ones on the fly.

**Review by generall — APPROVED** [review]

**Comment by agourlay** [comment]
@zzzz-vincent I decided to merge it as it is to not block you from making progress.
Keep up the good work :+1:

**Pull request merged**